RHDEVDOCS-6126 Document how to create aggregated cluster roles (#81594)
openshift/openshift-docs: modules/gitops-enabling-the-creation-of-aggregated-cluster-roles.adoc

// Module included in the following assembly:
//
// * declarative_clusterconfig/customizing-permissions-by-creating-aggregated-cluster-roles.adoc

:_mod-docs-content-type: PROCEDURE
[id="gitops-enabling-the-creation-of-aggregated-cluster-roles_{context}"]
= Enabling the creation of aggregated cluster roles

> To enable the creation of aggregated cluster roles for the Argo CD Application Controller component of a cluster-scoped Argo CD instance, you must configure the corresponding field by editing the YAML file of the Argo CD custom resource (CR).

Reviewer: Suggestion: Can we add the explicit name of the corresponding field in this sentence? As it is the first sentence of the procedure, I feel that providing the name of the corresponding field improves the existing sentence.

> 

Reviewer 2: missing prereqs

Author: Thanks a lot for the feedback. Please refer to the Assembly prerequisites section [here](https://redhat-documentation.github.io/modular-docs/#assembly-guidelines). The GitOps standalone documentation follows this method. To keep it consistent with what we already have, I prefer to retain this section as is.

.Procedure

> . In the Argo CD CR, set the value of the `.spec.defaultClusterScopedRoleDisabled` field to `true`:

Reviewer 3: `.spec.defaultClusterScopedRoleDisabled` => `.spec.aggregatedClusterRoles`

It is correct in example yaml below.

+
.Example Argo CD CR
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: example # <1>
  namespace: spring-petclinic # <2>
# ...
spec:
  aggregatedClusterRoles: true # <3>
# ...
----
<1> The name of the cluster-scoped instance.
<2> The namespace where you want to run the cluster-scoped instance.

> <3> The value set to `true` enables the creation of aggregated cluster roles. If you do not want to enable the creation of aggregated cluster roles, do not include this line or set the value to `false`.

Reviewer 2: ```suggestion
<3> The value set to `true` enables the creation of aggregated cluster roles. If you do not want to enable the creation of aggregated cluster roles, either do not include this line or set the value to `false`.
```
either/or
neither/nor
OK, I could only find a British citation, but it's the same in the US, https://dictionary.cambridge.org/us/grammar/british-grammar/either-or

Author: Addressed

+
.Example output
[source,terminal]
----
argocd.argoproj.io/example configured
----

. Verify that the `Status` field of the cluster-scoped Argo CD instance shows as `Phase: Available` by running the following command:
+
[source,terminal]
----
$ oc describe argocd.argoproj.io/example -n spring-petclinic
----
+
.Example output
[source,terminal]
----
Name:         example
Namespace:    spring-petclinic
Labels:       <none>
Annotations:  <none>
API Version:  argoproj.io/v1beta1
Kind:         ArgoCD
Metadata:
  Creation Timestamp:  2024-08-14T08:20:53Z
  Finalizers:
    argoproj.io/finalizer
  Generation:        3
  Resource Version:  60437
  UID:               57940e54-d60b-4c1a-bc4a-85c81c63ab69
Spec:

>   Aggregated Cluster Roles:  true

Reviewer 2: do we need this whole example? or could we just 
# ... 
here and then pick back up on line 102? because it's a looong scan to get down to the info we're looking for

Author: Addressed

  Controller:
    Processors:
    Sharding:
  Grafana:
    Enabled:  false
    Ingress:
      Enabled:  false
    Route:
      Enabled:  false
  Ha:
    Enabled:  false
  Initial SSH Known Hosts:
  Monitoring:
    Enabled:  false
  Notifications:
    Enabled:  false
  Prometheus:
    Enabled:  false
    Ingress:
      Enabled:  false
    Route:
      Enabled:  false
  Rbac:
  Redis:
  Repo:
  Server:
    Autoscale:
      Enabled:  false
    Grpc:
      Ingress:
        Enabled:  false
    Ingress:
      Enabled:  false
    Route:
      Enabled:  false
    Service:
      Type:  
  Tls:
    Ca:
Status:
  Application Controller:      Running
  Application Set Controller:  Unknown
  Phase:                       Available # <1>
  Redis:                       Running
  Repo:                        Running
  Server:                      Running
  Sso:                         Unknown
Events:                        <none>
----

> <1> The status indicates that the cluster-scoped Argo CD instance is healthy and available.

Reviewer 2: ```suggestion
<1> The `Available` status indicates that the cluster-scoped Argo CD instance is healthy and available.
```
I assume there is an unhealthy status.

Author: Addressed

+

> The {gitops-title} Operator creates the following default cluster roles and manages them:

Reviewer 2: ```suggestion
[NOTE]
====
The {gitops-title} Operator creates the following default cluster roles and manages them:

* `<argocd_name>-<argocd_namespace>-argocd-application-controller` aggregated cluster role
* `<argocd_name>-<argocd_namespace>-argocd-application-controller-view`
* `<argocd_name>-<argocd_namespace>-argocd-application-controller-admin` 
====
```
I think this info should be a NOTE because it isn't part of the procedure.

Author: Addressed

+

> * `<argocd_name>-<argocd_namespace>-argocd-application-controller` aggregated cluster role

Reviewer 2: ```suggestion
```

> * `<argocd_name>-<argocd_namespace>-argocd-application-controller` aggregated cluster role

Reviewer: Same feedback as the earlier comment above.

Author: Sorted the conversation with @Dhruv-Soni11. Retaining the content as is.

> * `<argocd_name>-<argocd_namespace>-argocd-application-controller-view`

Reviewer 2: ```suggestion
```

> * `<argocd_name>-<argocd_namespace>-argocd-application-controller-admin` 

Reviewer 2: ```suggestion
```

> . Verify that the Operator has created the default cluster roles and cluster role bindings for the Argo CD Application Controller and Argo CD server components by running the following commands:

Reviewer 2: ```suggestion
. Verify that the Operator created the default cluster roles and cluster role bindings for the Argo CD Application Controller and Argo CD server components by running the following commands:
```

+
[source,terminal]
----
$ oc get ClusterRoles -l app.kubernetes.io/part-of=argocd
----
+
.Example output
[source,terminal]
----
NAME                                                           CREATED AT
example-spring-petclinic-argocd-application-controller         2024-08-14T08:20:58Z
example-spring-petclinic-argocd-application-controller-admin   2024-08-14T09:08:38Z
example-spring-petclinic-argocd-application-controller-view    2024-08-14T09:08:38Z
example-spring-petclinic-argocd-server                         2024-08-14T08:20:59Z
----
+
[source,terminal]
----
$ oc get ClusterRoleBindings -l app.kubernetes.io/part-of=argocd
----
+
.Example output
[source,terminal]
----
NAME                                                     ROLE                                                                 AGE
example-spring-petclinic-argocd-application-controller   ClusterRole/example-spring-petclinic-argocd-application-controller   54m
example-spring-petclinic-argocd-server                   ClusterRole/example-spring-petclinic-argocd-server                   54m
----
+

> The cluster role bindings for the `view` and `admin` cluster roles are not created. This is because the cluster role bindings only add permissions to the aggregated cluster role and do not directly configure permissions for the Argo CD Application Controller.

Reviewer 3: LGTM, just a suggestion, we can rephrase this line, feel free to ignore if you disagree,.  

"This is because the cluster role bindings only add permissions to the aggregated cluster role and do not directly configure permissions for the Argo CD Application Controller." =>
"This is because they don’t directly configure permissions to the Argo CD Application Controller, instead they add permissions to aggregated cluster role which  has a cluster role binding."

Author: Thanks @jparsai. Addressed as discussed. Thanks!

+
[TIP]
====
Alternatively, you can use the {OCP} web console to verify from the *Administrator* perspective. You can go to *User Management* -> *Roles* and *User Management* -> *RoleBindings*, respectively. You can search for the cluster roles and cluster role bindings that have the `app.kubernetes.io/part-of:argocd` label.
====

> . Verify that the aggregated cluster role is created by checking the permissions of the aggregated cluster role,  `view`, and `admin` cluster roles by running the following command:

Reviewer 2: ```suggestion
. Verify that the aggregated cluster role is created by checking the permissions of outputs of the roles by running the following command:
```
That sentence felt like word salad to me; I would just simply. Details are already in each YAML set of footnotes. 
Does this one command output all three YAMLs?

Author: No. You will have to replace `<cluster_role_name>` with the name of the role you created and check it for each role.

+
[source,terminal]
----

> $ oc get ClusterRole/<cluster_role_name> -o yaml

Reviewer 2: ```suggestion
$ oc get ClusterRole/<cluster_role_name> -o yaml # <1>
```
assuming this is a user-replaced value, we need a footnote

Author: Addressed

----

> +

Reviewer 2: ```suggestion
<1> Replace `<cluster_role_name>` with the name of the role you created.
+
```

Author: Addressed

    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1beta1","kind":"ArgoCD","metadata":{"annotations":{},"name":"example","namespace":"spring-petclinic"},"spec":{"aggregatedClusterRoles":true}}
    rbac.authorization.kubernetes.io/autoupdate: "true"
  creationTimestamp: "2024-08-14T09:59:15Z"
  labels:
    app.kubernetes.io/managed-by: spring-petclinic
    app.kubernetes.io/name: example
    app.kubernetes.io/part-of: argocd
    argocd/aggregate-to-controller: "true"
  name: example-spring-petclinic-argocd-application-controller-admin # <1>
  resourceVersion: "78642"
  uid: e2d35b6f-0832-4993-8b24-915a725454f9
aggregationRule: # <2>
  clusterRoleSelectors:
  - matchLabels:
      app.kubernetes.io/managed-by: spring-petclinic
      argocd/aggregate-to-admin: "true"
rules: null # <3>
----
<1> The name of the `admin` cluster role.
<2> The predefined list of labels indicates that the existing `<argocd_name>-<argocd_namespace>-argocd-application-controller-admin` cluster role can inherit permissions from the other user-defined cluster roles.
<3> Specifies that no permissions are defined yet in one or more user-defined cluster roles.
+
[TIP]
====

> Alternatively, you can use the {OCP} web console to verify from the *Administrator* perspective. You can go to *User Management* -> *Roles*, use the *Filter* option, select *Cluster-wide Roles*, and search for the aggregated cluster role, `view`, and `admin` cluster roles. You have to open the cluster role to check the details and configurations.

Author: Addressed

====
+

> As a cluster administrator, you can now create one or more user-defined cluster roles and configure user-defined permissions for Argo CD Application Controller.

Author: Thanks for the inputs. We are saving this for our Content journey and content improvement work. For now, we cannot implement this for this chapter alone.

To keep it consistent with what we already have in the GitOps docs, I prefer to retain this info as is. However, we will definitely implement this in our future doc versions. Thanks again :)
